stats/lib/request-helpers.js

```diff
@@ -89,7 +89,7 @@ export const getStatsWithFilterAndCaching = async (pathname, pathParams, searchP
 const setCacheControlForStatsResponse = (res, filter) => {
   // We cannot simply compare filter.to vs today() because there may be a delay in finalizing
   // stats for the previous day. Let's allow up to one hour for the finalization.
-  const boundary = getDayAsISOString(new Date(Date.now() - 3600_000))
+  const boundary = getDayAsISOString(new Date(todayDate().getTime() - 3600_000))
```

Comment:
Oh, I see now why you need `todayDate`.

Let's find more descriptive names for `todayDate` and `getDayAsISOString` then. The names should convey that:

- `todayDate` is a date-time value which, when converted to UTC timezone, produces the same date (day) as is in the local timezone now.
- `getDayAsISOString` returns the date part of the provided parameter expressed in the UTC timezone

That sounds too complex to me. Maybe we are approaching this from a wrong angle.

What we want to implement:
- If we are inside the interval `0:00:00-1:00:00` in the local timezone, then we compare `filter.to` against _yesterday_ date.
- Otherwise, we compare `filter.to` against _today_ date.

How about converting `filter.to` to a full date-time value (in local timezone) and the comparing it against `new Date()`?

```js
// Convert "2024-09-01" to "2024-09-02T01:00:00.000" in local timezone
const adjustedFilterTo = new Date(
  new Date(filter.to) + 25 * 3600_000 
)
if (adjustedFilterTo >= new Date()) {
  // response includes partial data for today, cache it for 10 minutes only
// etc.
```

Reply:
> todayDate is a date-time value which, when converted to UTC timezone, produces the same date (day) as is in the local timezone now.

I think this is not true, unless I am misunderstanding. `todayDate` is simply the local date based on the timezone of the computer running the code, as far I understand it. Maybe we can change the name to `todayLocalDate`?

> getDayAsISOString returns the date part of the provided parameter expressed in the UTC timezone

I think this name is fine? Its functionality is unchanged from before in the current form of this PR, and describes its behavior accurately I think. It does not have any behavior related to UTC timezone, not sure if I am misunderstanding your statement here.

> How about converting filter.to to a full date-time value (in local timezone) and the comparing it against new Date()?

I don't know, to me this seems harder to understand than the PR as it is now. But I'll default to whatever you and/or Julian decide on.

Reply:
I think the current solution have so many subtle nuances and complexities that we should step back and look for better options.

What we are trying to achieve: Given a `Date` value in the local time zone, we want to get a date string in the YYYY-MM-DD format for the local day.

How about this solution:

- Create a function called `getLocalDayAsISOString`, to make it clear this function is preserving the local timezone
- Implement the function by calling year/month/day getters, padding the values to 2/4 digits and concatenating them to the desired format.

```js
const getLocalDayAsISOString = (d) => {                                         
  return [                                                                      
    d.getFullYear(),                                                            
    String(d.getMonth() + 1).padStart(2, '0'),                                  
    String(d.getDate()).padStart(2, '0')                                        
  ].join('-')                                                                   
}
```

This function is very explicit about what it does. I think it should also not require any changes in how we calculate the values passed in the `d` parameter.

At the top of this file:

```js
export const today = () => getLocalDayAsISOString(new Date())
export const yesterday = () => getLocalDayAsISOString(new Date(Date.now() - 24 * 60 * 60 * 1000))
```

In the code we are discussing in this thread:

```js
const boundary = getLocalDayAsISOString(new Date(Date.now() - 3600_000))
```

@juliangruber @PatrickNercessian WDYT?

Reply:
+1 to being very explicit here. The previous version to me also is too magic.

Could the proposed version be simplified to this?

```js
const getLocalDayAsISOString = (d) => {
  return d.toISOString().split('T')[0]
}
```

I believe we had this version before, so I think this is _not_ what we need, which also tells me that the version @bajtos proposed isn't going to work. @PatrickNercessian what are we missing?

Reply:
> @juliangruber I think the issue with that one is that it is does not preserve the local date, which was the original problem.

Right, `d.toISOString()` is in UTC.

@bajtos's version works, as Date methods like `getDay()` return according to local time (see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getDay).

I'm in favor of using that. And I'm happy I learned something new about JS Date functions

Reply:
I feel we are spending too much time on a very minor issue. What's the simplest way to resolve this?

Here are two options I like:

1. We can rule that **the DB must run on a computer configured with the UTC timezone**. This matches our production system, GitHub Actions, and some Docker deployments of Postgres. In this case, I believe the original implementation `d.toISOString().split('T')[0]` works best. However, this condition is not met when the DB runs as a local service on the dev computer - @PatrickNercessian's case.

2. Alternatively, we can rule that **the DB must run in the same timezone as the Node.js processes**. This matches our production system, GitHub Actions, and a database server running as a local service. In this case, I believe the string-manipulation based solution I proposed in https://github.com/filecoin-station/spark-stats/pull/210#discussion_r1756908718 works best. However, this condition is not met when the DB runs via Docker on the dev computer - @bajtos's case.

What happens when the condition is not met: there is a window around midnight UTC when the tests fail when executed locally. 

IMO, the second option - "DB and Node.js processes have the same timezone" - is what most people would assume, and it's also the easiest to reason about. Therefore, I prefer to go with that.

Personally, I will either figure out how to run the DB in my local timezone, or, and this is more likely, I'll remember that the tests will fail when I run them between midnight and 2am CEST, and ignore the problem for now.

> All of this depends on the postgres configuration, so I think we must do an SQL query for this. If we switch javascript to use local date, and PostGres is set to UTC (as in your case), tests will also fail at certain times.

This adds an extra query to our REST API endpoints. IIUC, its only benefit is solving an edge case on developer machines only. I'd rather avoid paying that performance cost.

Reply:
Gotcha. Made the change to use the string-manipulation solution you suggested. However, the downside it's duplicated logic, in both `stats` and `observer`, since they both require the functionality.

We could always put them in `db`, such as what we did with the test-helper `givenDailyParticipants`, but this seems even less related to the database. And it's also not a test-helper really, since it's also used in the production code.

So I ended up duplicating them. Let me know if you'd prefer some other solution

Reply:
To me this method isn't big enough to require deduplication

Reply:
I am fine with duplication, too.